File: app/src/main/java/com/chaitupenjudcoder/firebasehelpers/FirebaseCategoriesHelper.java
package com.chaitupenjudcoder.firebasehelpers;

import android.support.annotation.NonNull;

import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;

import java.util.ArrayList;

//This class helps retrieve categories from Firebase Database and increases code resulability
public class FirebaseCategoriesHelper {

    private DatabaseReference mCategoryRef;
    private FirebaseAuth mAuth;
    private FirebaseUser mUser;
    private String uid;
    private ArrayList<String> categoriesList = new ArrayList<>();
    private DatabaseReference totalsRef;

    //  This interface is called after creating object of this class to use as callback in other
    public interface GetAllCategory {
        void categories(ArrayList<String> categoriesList);
    }

    public interface GetCategoryTotal {
        void categoriesTotal(int total);
    }

    //  Constructor to instantiate all firebase variables
    public FirebaseCategoriesHelper() {
        mAuth = FirebaseAuth.getInstance();
        mUser = mAuth.getCurrentUser();
        uid = mUser.getUid();
        mCategoryRef = FirebaseDatabase.getInstance().getReference("data/" + uid + "/categories");
        //  set the keep synced true for offline capability
        mCategoryRef.keepSynced(true);
    }

    public void getCategoryTotal(final GetCategoryTotal getTotals, final String category) {
        totalsRef = FirebaseDatabase.getInstance().getReference("data/" + uid + "/spendings");

        ValueEventListener totalListener = new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                String bucksStr;
                int incExpTotal = 0;
                for (DataSnapshot incomeExpenseTotal : dataSnapshot.getChildren()) {
                    bucksStr = incomeExpenseTotal.child("bucksString").getValue(String.class);
                    int amount = Integer.valueOf(incomeExpenseTotal.child("amount").getValue(String.class));
                    if (bucksStr != null) {
                        if (bucksStr.equals(category)) {
                            incExpTotal += amount;
                        }
                    }
                }
                getTotals.categoriesTotal(incExpTotal);
            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {

            }
        };
        totalsRef.addValueEventListener(totalListener);
    }

    //  This method is called with this class' object everywhere and override the interface defined
    public void getAllCategories(final GetAllCategory allCategory, final String BUCKS_STRING) {
        ValueEventListener categoryListener = new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                categoriesList.clear();
                for (DataSnapshot categorieShot : dataSnapshot.child(BUCKS_STRING).getChildren()) {
                    String category = categorieShot.getValue(String.class);
                    categoriesList.add(category);
                }
                allCategory.categories(categoriesList);
            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {

            }
        };
        mCategoryRef.addValueEventListener(categoryListener);
    }
}

File: app/src/main/java/com/chaitupenjudcoder/firebasehelpers/FirebaseCategoriesHelper.kt
package com.chaitupenjudcoder.firebasehelpers

import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.*
import kotlinx.coroutines.tasks.await
import java.util.*

//This class helps retrieve categories from Firebase Database and increases code re-usability
class FirebaseCategoriesHelper {
    private val mCategoryRef: DatabaseReference
    private val mAuth get() = FirebaseAuth.getInstance()
    private val mUser get() = mAuth.currentUser
    private val uid get() = mUser!!.uid
    private val categoriesList = ArrayList<String>()
    private var totalsRef: DatabaseReference? = null

    //  Constructor to instantiate all firebase variables
    init {
        mCategoryRef = FirebaseDatabase.getInstance().getReference("data/$uid/categories")
        //  set the keep synced true for offline capability
        mCategoryRef.keepSynced(true)
    }

    suspend fun getCategoryTotal(category: String, categoriesTotal: suspend (total: Int) -> Unit) {
        val totalsSnapshot = FirebaseDatabase.getInstance().getReference("data/$uid/spendings").get().await()
        var bucksStr: String
        var incExpTotal = 0

        for (incomeExpenseTotal in totalsSnapshot.children) {
            bucksStr = incomeExpenseTotal.child("bucksString").getValue(
                String::class.java
            ).toString()
            val amount = Integer.valueOf(
                incomeExpenseTotal.child("amount").getValue(
                    String::class.java
                ).toString()
            )
            if (bucksStr == category) {
                incExpTotal += amount
            }
        }

        categoriesTotal(incExpTotal)
    }

    //  This method is called with this class' object everywhere and override the interface defined
    suspend fun getAllCategories(BUCKS_STRING: String?, categories: suspend (ArrayList<String>) -> Unit) {
        val categorySnapshot = mCategoryRef.get().await()

        categoriesList.clear()
        for (categoryShot in categorySnapshot.child(BUCKS_STRING!!).children) {
            val category = categoryShot.getValue(String::class.java)
            categoriesList.add(category!!)
        }

        categories(categoriesList)
    }
}